### PR TITLE
Support metrics counter types in ESQL

### PR DESCRIPTION
**Pull request by dnhatn**
This commit adds support for numeric metrics counter fields in ES|QL. These counter types, including `counter_long`, `counter_integer`, and `counter_double`, are different from their parent types. Users will have limited interaction with these counter types, restricted to:

- Retrieving values without any processing
- Casting to their root type (e.g., ``to_long(a_long_counter)``)
- Using them in the metrics rate aggregation

These restrictions are intentional to prevent misuse. If users want to use them as numeric values, explicit casting to their root types is required.

### CODE REVIEW

**Review by astefan — COMMENTED** [review]
Looks good in general, but there are edge cases that might need some refining:
- `stats min(tx)` -> `argument of [min(tx)] must be [datetime or numeric except unsigned_long], found value [min(tx)] type [counter_long]`. It's good we have a verification exception, but its message is not entirely true
- tests are needed in VerifierTests for error messages when using these types in various places in queries
- it seems I cannot aggregate on these types, but I can group by them (ie `min(time) by counter`). Is this expected?
- `count(counter)` seems to also work...

File: docs/reference/esql/functions/kibana/definition/to_long.json
Comment (nik9000) [inline docs/reference/esql/functions/kibana/definition/to_long.json:23]: @drewdaemon this is fun - did you know that things annotated with `counter` aren't `long` to ESQL? Before this PR they were unsupported entirely. With this PR they'll have types like this.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDouble.java
Comment (nik9000) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDouble.java:47]: Should we allow `counter_int` and `counter_long` in all of these too?

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/LocalExecutionPlanner.java
Comment (nik9000) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/LocalExecutionPlanner.java:347]: This'd allow us to sort on these. Do we want that?

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypes.java
Comment (nik9000) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypes.java:56]: Probably worth a javadoc for folks who haven't been following along.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypes.java
Comment (nik9000) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypes.java:223]: I think counters are representable though. They can be loaded into blocks now.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/action/EsqlQueryResponseTests.java
Comment (nik9000) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/action/EsqlQueryResponseTests.java:132]: I always get this backwards - can the `type` include a counter type here? I think it should because we can return them.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java
Comment (nik9000) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java:994]: I've grown to hate this `Map`. I made it, but I dislike it....

Comment (nik9000) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java:994]: You've done the right thing to it, but I'm sorry you had to.

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/40_tsdb.yml
Comment (nik9000) [reply x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/40_tsdb.yml:161]: :+1: on rename.

**Review by nik9000 — COMMENTED** [review]
Left some questions. Probably good to go but I'll recheck after you answer.

File: docs/reference/esql/functions/kibana/definition/to_long.json
Comment (drewdaemon) [reply docs/reference/esql/functions/kibana/definition/to_long.json:23]: On our side, we use Javascript types so it's all `number` :D 

But that may change so this is very good to keep in mind. Thanks for the ping (cross ref with https://github.com/elastic/kibana/issues/174710)

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/40_tsdb.yml
Comment (dnhatn) [reply x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/40_tsdb.yml:161]: ++, renamed in https://github.com/elastic/elasticsearch/pull/107877/commits/0222b149aa4078f95d7c898a09813c84d23b6732

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/LocalExecutionPlanner.java
Comment (dnhatn) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/LocalExecutionPlanner.java]: Definitely not, my bad. I fixed this in https://github.com/elastic/elasticsearch/pull/107877/commits/06b30783b80286edfbf80f48619ec7e0fca0e714.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypes.java
Comment (dnhatn) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypes.java]: ++, added in https://github.com/elastic/elasticsearch/pull/107877/commits/472fbd2de7b86babfbf5076395b0519ac7687595

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/action/EsqlQueryResponseTests.java
Comment (dnhatn) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/action/EsqlQueryResponseTests.java]: I removed this restriction in https://github.com/elastic/elasticsearch/pull/107877/commits/5c2b08fbbb180065b34ae4ea50cb377fafe16712

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypes.java
Comment (dnhatn) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypes.java:230]: I disabled this mainly to avoid the map in AbstractFunctionTestCase https://github.com/elastic/elasticsearch/pull/107877#discussion_r1579527982. Would you be okay if I enable this in a follow-up?

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDouble.java
Comment (dnhatn) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDouble.java:47]: I intended to minimize support for these types, but I'm fine with either way. What's your preference?

**Comment by dnhatn** [comment]
> count(counter) seems to also work...

> Should we allow counter_int and counter_long in all of these too?

Good question. Some usages, such as count(counter) or to_string(counter), seem to make sense. Should we support them all, or still require casting in these cases?

**Comment by dnhatn** [comment]
@astefan I pushed https://github.com/elastic/elasticsearch/pull/107877/commits/5b09bdfeaed3749f920b1c1b050ccbe9a3482136 to add verification tests and restrict usages in count, aggregation, and grouping.

**Comment by astefan** [comment]
> > count(counter) seems to also work...
> 
> > Should we allow counter_int and counter_long in all of these too?
> 
> Good question. Some usages, such as count(counter) or to_string(counter), seem to make sense. Should we support them all, or still require casting in these cases?

I would be cautious and start with requiring casting everywhere, there may be edge cases that we couldn't think of atm and it may  be difficult to remove the support once it's merged. If we get demand for this, we'll analyze these requests and adapt at that time.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypes.java
Comment (nik9000) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypes.java:230]: :+1:

Delaying is fine.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java
Comment (nik9000) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java:994]: There's another version of `errorsForCasesWithoutExamples` which takes another parameter that outputs a matcher for hte expected type. You could use that rather than changing the map.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/NumericAggregate.java
Comment (costin) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/NumericAggregate.java:43]: Nit: `except [unsigned_long, counter_*(double, integer, long)]`

Compacted the exception message a bit.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDouble.java
Comment (costin) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDouble.java:47]: For consistency it makes sense to add them.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/LocalExecutionPlanner.java
Comment (costin) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/LocalExecutionPlanner.java:347]: Separate from this PR but it makes sense to centralize these strings in one class to avoid bugs due to typos.

**Review by costin — COMMENTED** [review]
This looks Nhat. Outside some superficial nits, the main comment I have is around CSV tests with counters in various places such as:
- filtering (WHERE counter > 10)
- expressions both named and anonymous  + scalar functions (EVAL X = counter % 3 > 2 | WHERE counter_long > counter_double)
- using it in agg functions MAX(counter), MIN(counter % 2)

Thanks!

**Review by astefan — APPROVED** [review]
Apart from Costin's comments (especially the additional tests), it's LGTM.

**Comment by nik9000** [comment]
> Outside some superficial nits, the main comment I have is around CSV tests with counters in various places such as:
> 
>     * filtering (WHERE counter > 10)
> 
>     * expressions both named and anonymous  + scalar functions (EVAL X = counter % 3 > 2 | WHERE counter_long > counter_double)
> 
>     * using it in agg functions MAX(counter), MIN(counter % 2)

I *think* all of those should fail unless you wrap them in `to_integer` or whatever. Which we may as well test paranoia. Tests are cheap and run fast and make sure future me doesn't break them. Future me is a menace.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/NumericAggregate.java
Comment (dnhatn) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/NumericAggregate.java]: ++, I pushed https://github.com/elastic/elasticsearch/pull/107877/commits/165385ef238809644ffcb92228afce0ea1bfdaa4

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDouble.java
Comment (dnhatn) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDouble.java]: ok, I added these in https://github.com/elastic/elasticsearch/pull/107877/commits/84733ff2060dec7600d94fde075432d6522d3433.

**Comment by dnhatn** [comment]
Thank you all for your reviews. I'll have to delay the CSV tests because I'm going to build a CSV dataset for the tsdb suite. Also, the CSV test infra doesn't support error checking yet. The YAML tests should have good coverage.

**Pull request merged**
